src/auth.rs

```diff
@@ -90,48 +95,59 @@ impl Middleware<crate::MiniserveConfig> for Auth {
         req: &HttpRequest<crate::MiniserveConfig>,
         resp: HttpResponse,
     ) -> Result<Response> {
-        if let Some(ref required_auth) = req.state().auth {
-            if let Some(auth_headers) = req.headers().get(header::AUTHORIZATION) {
-                let auth_req = match parse_basic_auth(auth_headers) {
-                    Ok(auth_req) => auth_req,
-                    Err(err) => {
-                        let auth_err = ContextualError::HTTPAuthenticationError(Box::new(err));
-                        return Ok(Response::Done(HttpResponse::BadRequest().body(
+        let required_auth = &req.state().auth;
+
+        if required_auth.is_empty() {
+            return Ok(Response::Done(resp));
+        }
+
+        if let Some(auth_headers) = req.headers().get(header::AUTHORIZATION) {
+            let auth_req = match parse_basic_auth(auth_headers) {
+                Ok(auth_req) => auth_req,
+                Err(err) => {
+                    let auth_err = ContextualError::HTTPAuthenticationError(Box::new(err));
+                    return Ok(Response::Done(
+                        HttpResponse::BadRequest().body(
                             build_unauthorized_response(
                                 &req,
                                 auth_err,
                                 true,
                                 StatusCode::BAD_REQUEST,
                             ),
-                        )));
-                    }
-                };
-                if !match_auth(auth_req, required_auth) {
-                    return Ok(Response::Done(HttpResponse::Unauthorized().body(
-                        build_unauthorized_response(
-                            &req,
-                            ContextualError::InvalidHTTPCredentials,
-                            true,
-                            StatusCode::UNAUTHORIZED,
                         ),
-                    )));
+                    ));
                 }
-            } else {
-                let new_resp = HttpResponse::Unauthorized()
-                    .header(
-                        header::WWW_AUTHENTICATE,
-                        header::HeaderValue::from_static("Basic realm=\"miniserve\""),
-                    )
-                    .body(build_unauthorized_response(
+            };
+
+            if match_auth(auth_req, required_auth) {
+                return Ok(Response::Done(resp));
+            }
+
+            return Ok(Response::Done(
+                HttpResponse::Unauthorized().body(
+                    build_unauthorized_response(
```

Comment:
I know we didn't do this before but might as well change it now: What do you think about sending `WWW_AUTHENTICATE` here too to give users an easier time after mistyping?

Reply:
I am not familiar with HTTP, what `WWW_AUTHENTICATE` does exactly?